src/java/org/apache/cassandra/net/MessageDelivery.java

```diff
@@ -74,9 +85,162 @@ public void onFailure(InetAddressAndPort from, RequestFailureReason reason)
+                                                                   TriFunction<Integer, InetAddressAndPort, RequestFailureReason, Boolean> shouldRetry,
+                                                                   RetryErrorMessage errorMessage)
+    {
+        return sendWithRetries(new AsyncPromise<>(), (Integer i, Message<RSP> msg) -> msg, backoff, retryThreads, verb, request, Iterators.cycle(candidate), shouldRetry, errorMessage);
+    }
+
+    public default <REQ, MSG_RSP, RSP> Future<RSP> sendWithRetries(BiFunction<Integer, Message<MSG_RSP>, RSP> msgToRsp,
+                                                                   Backoff backoff,
+                                                                   RetryScheduler retryThreads,
+                                                                   Verb verb, REQ request,
+                                                                   InetAddressAndPort candidate,
+                                                                   TriFunction<Integer, InetAddressAndPort, RequestFailureReason, Boolean> shouldRetry,
+                                                                   RetryErrorMessage errorMessage)
+    {
+        return sendWithRetries(new AsyncPromise<>(), msgToRsp, backoff, retryThreads, verb, request, Iterators.cycle(candidate), shouldRetry, errorMessage);
+    }
+
+    public default <REQ, MSG_RSP, RSP> Future<RSP> sendWithRetries(Promise<RSP> promise,
+                                                                   BiFunction<Integer, Message<MSG_RSP>, RSP> msgToRsp,
+                                                                   Backoff backoff,
+                                                                   RetryScheduler retryThreads,
+                                                                   Verb verb, REQ request,
+                                                                   Iterator<InetAddressAndPort> candidates,
+                                                                   TriFunction<Integer, InetAddressAndPort, RequestFailureReason, Boolean> shouldRetry,
+                                                                   RetryErrorMessage errorMessage)
+    {
+        sendWithRetries(this, promise, msgToRsp, backoff, retryThreads, verb, request, candidates, shouldRetry, errorMessage, 0);
+        return promise;
+    }
     public <V> void respond(V response, Message<?> message);
     public default void respondWithFailure(RequestFailureReason reason, Message<?> message)
     {
         send(Message.failureResponse(message.id(), message.expiresAtNanos(), reason), message.respondTo());
     }
+
+    interface RetryErrorMessage
+    {
+        String apply(int attempt, ResponseFailureReason retryFailure, @Nullable InetAddressAndPort from, @Nullable RequestFailureReason reason);
+    }
+
+    private static <REQ, MSG_RSP, RSP> void sendWithRetries(MessageDelivery messaging,
```

Comment:
What if we tried something like this:

```
    interface OnResult<T>
    {
        public void result(int attempt, Message<T> success, Throwable failure);
    }

    interface RetryErrorMessage
    {
        String apply(int attempt, ResponseFailureReason retryFailure, @Nullable InetAddressAndPort from, @Nullable RequestFailureReason reason);
    }

    private static <REQ, MSG_RSP> void sendWithRetries(MessageDelivery messaging,
                                                       OnResult<REQ> onResult,
                                                       Backoff backoff,
                                                       RetryScheduler retryThreads,
                                                       Verb verb, REQ request,
                                                       Iterator<InetAddressAndPort> candidates,
                                                       TriFunction<Integer, InetAddressAndPort, RequestFailureReason, Boolean> shouldRetry,
                                                       RetryErrorMessage errorMessage,
                                                       int attempt)

    {
        Promise<REQ> promise = new AsyncPromise<>();
        sendWithRetries(messaging, 
                        (att, res, throwable) -> {
                            if (res != null)
                                promise.setSuccess(res.payload);
                            else 
                                promise.setFailure(throwable);
                        },
                        backoff,
                        retryThreads,
                        verb, 
                        request,
                        candidates, 
                        shouldRetry, 
                        errorMessage,
                        0);
        try
        {
            promise.await().get();
        }
        catch (Throwable e)
        {
            throw new RuntimeException(e);
        }
    }
    private static <REQ, RSP> void sendWithRetriesInternal(MessageDelivery messaging,
                                                           OnResult<RSP> onResult,
                                                           Backoff backoff,
                                                           RetryScheduler retryThreads,
                                                           Verb verb, REQ request,
                                                           Iterator<InetAddressAndPort> candidates,
                                                           TriFunction<Integer, InetAddressAndPort, RequestFailureReason, Boolean> shouldRetry,
                                                           RetryErrorMessage errorMessage,
                                                           int attempt)
    {
        class Request implements RequestCallbackWithFailure<RSP>
        {
            @Override
            public void onResponse(Message<RSP> msg)
            {
                onResult.result(attempt, msg, null);
            }

            @Override
            public void onFailure(InetAddressAndPort from, RequestFailureReason failure)
            {
                if (Thread.currentThread().isInterrupted())
                {
                    onResult.result(attempt, null, new InterruptedException(errorMessage.apply(attempt, ResponseFailureReason.Interrupted, null, null)));
                    return;
                }
                if (!candidates.hasNext())
                {
                    onResult.result(attempt, null, new NoMoreCandidatesException(errorMessage.apply(attempt, ResponseFailureReason.NoMoreCandidates, null, null)));
                    return;
                }
                if (!backoff.mayRetry(attempt))
                {
                    onResult.result(attempt, null, new MaxRetriesException(attempt, errorMessage.apply(attempt, ResponseFailureReason.MaxRetries, from, failure)));
                    return;
                }
                if (!shouldRetry.apply(attempt, from, failure))
                {
                    onResult.result(attempt, null, new FailedResponseException(from, failure, errorMessage.apply(attempt, ResponseFailureReason.Rejected, from, failure)));
                    return;
                }
                try
                {
                    retryThreads.schedule(() -> sendWithRetries(messaging, onResult, backoff, retryThreads, verb, request, candidates, shouldRetry, errorMessage, attempt + 1),
                                          backoff.computeWaitTime(attempt), backoff.unit());
                }
                catch (Throwable t)
                {
                    onResult.result(attempt, null, new FailedScheduleException(errorMessage.apply(attempt, ResponseFailureReason.FailedSchedule, from, failure), t));
                }
            }
        }
        messaging.sendWithCallback(Message.outWithFlag(verb, request, CALL_BACK_ON_FAILURE), candidates.next(), new Request());
    }
```

Reply:
why?  I am cool with the general idea... if you don't want to allocate a `Promise` we don't force you at the lowest level API... just making sure I understand where you are coming from

Reply:
one thing we loose is `if (promise.isDone() || promise.isCancelled()) return;`.  this isn't really a big deal as we *own* the promise... but if one is provided (think it is in TCM) then we will keep retrying even if the caller doesn't care anymore

Reply:
for the TCM case can always add `if (promise.isDone() || promise.isCancelled()) return false;` into the retry function...

Reply:
pushed

Reply:
i did not have a strong opinion; was just trying to suggest an API that might be slightly easier to approach.